ir/instr.cpp

```diff
@@ -2373,48 +2373,109 @@ unique_ptr<Instr> Return::dup(const string &suffix) const {
 }
 
 
+Assume::Assume(Value &cond, Kind kind)
+    : Instr(Type::voidTy, "assume"), args({&cond}), kind(kind) {
+  assert(kind == AndNonPoison || kind == IfNonPoison || kind == WellDefined);
+}
+
+Assume::Assume(vector<Value *> &&args0, Kind kind)
+    : Instr(Type::voidTy, "assume"), args(move(args0)), kind(kind) {
+  if (args.size() == 1)
+    assert(kind == AndNonPoison || kind == IfNonPoison || kind == WellDefined);
+  else {
+    auto sz = args.size();
+    assert((kind == Align && sz == 2) ||
+           (kind == Range && sz > 1 && sz % 2 == 1));
+  }
+}
+
 vector<Value*> Assume::operands() const {
-  return { cond };
+  return args;
 }
 
 void Assume::rauw(const Value &what, Value &with) {
-  RAUW(cond);
+  for (auto &arg: args)
+    RAUW(arg);
 }
 
 void Assume::print(ostream &os) const {
   switch (kind) {
-  case AndNonPoison: os << "assume "; break;
-  case IfNonPoison: os << "assume_non_poison "; break;
-  case WellDefined: os << "assume_welldefined "; break;
+  case AndNonPoison: os << "assume"; break;
+  case IfNonPoison: os << "assume_non_poison"; break;
+  case WellDefined: os << "assume_welldefined"; break;
+  case Align: os << "assume_align"; break;
+  case Range: os << "assume_range"; break;
   }
-  os << *cond;
+  for (auto &arg: args)
+    os << ' ' << *arg;
 }
 
 StateValue Assume::toSMT(State &s) const {
   switch (kind) {
   case AndNonPoison: {
-    auto &v = s.getAndAddPoisonUB(*cond);
+    auto &v = s.getAndAddPoisonUB(*args[0]);
     s.addUB(v.value != 0);
     break;
   }
   case IfNonPoison: {
-    auto &[v, np] = s[*cond];
+    auto &[v, np] = s[*args[0]];
     s.addUB(np.implies(v != 0));
     break;
   }
   case WellDefined:
-    (void)s.getAndAddPoisonUB(*cond, true);
+    (void)s.getAndAddPoisonUB(*args[0], true);
+    break;
+  case Align: {
+    // assume(ptr, align)
+    const auto &vptr = s.getAndAddPoisonUB(*args[0]);
```

Comment:
why is it UB to pass a poison ptr? is this explicit in LangRef?
An assume bundle with align is often used after inlining a function that had an align arg. So the semantics needs to be the same. i.e. calling f(align 4 poison) can only be UB iff assume align 4 poison is UB.
We need some investigation to decide on the semantics.